pkg/parse/events/funnel_test.go

```diff
@@ -0,0 +1,256 @@
+					Result: Result{},
+				},
+			},
+		},
+		{
+			name: "ResyncEvents From SyncWithReimportPeriod",
+			builder: &PublishingGroupBuilder{
+				SyncWithReimportPeriod: time.Second,
+			},
+			stepSize: time.Second,
+			expectedEvents: []eventResult{
+				{
+					Event:  Event{Type: SyncWithReimportEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: SyncWithReimportEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: SyncWithReimportEventType},
+					Result: Result{},
+				},
+			},
+		},
+		{
+			name: "StatusEvents From StatusUpdatePeriod",
+			builder: &PublishingGroupBuilder{
+				StatusUpdatePeriod: time.Second,
+			},
+			stepSize: time.Second,
+			expectedEvents: []eventResult{
+				{
+					Event:  Event{Type: StatusEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: StatusEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: StatusEventType},
+					Result: Result{},
+				},
+			},
+		},
+		{
+			name: "NamespaceResyncEvents From NamespaceControllerPeriod",
+			builder: &PublishingGroupBuilder{
+				NamespaceControllerPeriod: time.Second,
+			},
+			stepSize: time.Second,
+			expectedEvents: []eventResult{
+				{
+					Event:  Event{Type: NamespaceResyncEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: NamespaceResyncEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: NamespaceResyncEventType},
+					Result: Result{},
+				},
+			},
+		},
+		{
+			name: "RetryEvents From RetryBackoff",
+			builder: &PublishingGroupBuilder{
+				RetryBackoff: wait.Backoff{
+					Duration: time.Second,
+					Factor:   0, // no backoff, just 1s delay
+					Steps:    3, // at least as many as we expect
+				},
+			},
+			stepSize: time.Second,
+			expectedEvents: []eventResult{
+				{
+					Event:  Event{Type: RetrySyncEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: RetrySyncEventType},
+					Result: Result{},
+				},
+				{
+					Event:  Event{Type: RetrySyncEventType},
+					Result: Result{},
+				},
+			},
+		},
+		{
+			name: "RetryEvents From RetryBackoff with ResetRetryBackoff",
+			builder: &PublishingGroupBuilder{
+				RetryBackoff: wait.Backoff{
+					Duration: time.Second,
+					Factor:   0, // no backoff, just 1s delay
+					Steps:    1, // just one, to show the count resets when ResetRetryBackoff is true
+				},
+			},
+			stepSize: time.Second,
+			expectedEvents: []eventResult{
+				{
+					Event: Event{Type: RetrySyncEventType},
+					Result: Result{
+						ResetRetryBackoff: true,
+					},
+				},
+				{
+					Event: Event{Type: RetrySyncEventType},
+					Result: Result{
+						ResetRetryBackoff: true,
+					},
+				},
+				{
+					Event: Event{Type: RetrySyncEventType},
+					Result: Result{
+						ResetRetryBackoff: true,
+					},
+				},
+			},
+		},
+		// TODO: test retry backoff behavior
+		// TODO: test all handlers together (can cause race conditions)
```

Comment:
It shouldn't. I just meant that with two timers sharing a clock you don't always get deterministicly ordered events, which makes it hard to compare against an expected result. 

It should be possible with the fake clock tho.

Reply:
I figured out how to test more of them together (SyncEvents, StatusEventType, ResyncEvents), but the retry and namespace events are very spammy, so I left them out. They were already spammy before this PR tho. So I think it's fine to leave them untested for now. Hopefully we can "fix" them to be less spammy in the future with the new event system, perhaps by triggering them with a channel, instead of a timer.

Reply:
Also added some backoff tests here too.